kernel/src/process.rs

```diff
@@ -119,127 +119,137 @@ impl fmt::Debug for ProcessLoadError {
 /// is able to create its own `load_processes()` function and use that instead.
 ///
 /// Processes are found in flash starting from the given address and iterating
-/// through Tock Binary Format headers. Processes are given memory out of the
-/// `app_memory` buffer until either the memory is exhausted or the allocated
-/// number of processes are created, with process structures placed in the
-/// provided array. How process faults are handled by the kernel is also
-/// selected.
+/// through Tock Binary Format (TBF) headers. Processes are given memory out of
+/// the `app_memory` buffer until either the memory is exhausted or the
+/// allocated number of processes are created. A reference to each process is
+/// stored in the provided `procs` array. How process faults are handled by the
+/// kernel must be provided and is assigned to every created process.
+///
+/// This function is made `pub` so that board files can use it, but loading
+/// processes from slices of flash an memory is fundamentally unsafe. Therefore,
+/// we require the `ProcessManagementCapability` to call this function.
+///
+/// Returns `Ok(())` if process discovery went as expected. Returns a
+/// `ProcessLoadError` if something goes wrong during TBF parsing or process
+/// creation.
 pub fn load_processes<C: Chip>(
     kernel: &'static Kernel,
     chip: &'static C,
     app_flash: &'static [u8],
-    app_memory: &mut [u8],
+    app_memory: &'static mut [u8],
     procs: &'static mut [Option<&'static dyn ProcessType>],
     fault_response: FaultResponse,
     _capability: &dyn ProcessManagementCapability,
 ) -> Result<(), ProcessLoadError> {
-    let mut remaining_flash = app_flash;
-    let mut app_memory_ptr = app_memory.as_mut_ptr();
-    let mut app_memory_size = app_memory.len();
-
     if config::CONFIG.debug_load_processes {
         debug!(
-            "Loading processes from flash={:#010X} into sram=[{:#010X}:{:#010X}]",
+            "Loading processes from flash=[{:#010X}:{:#010X}] into sram=[{:#010X}:{:#010X}]",
             app_flash.as_ptr() as usize,
-            app_memory_ptr as usize,
-            app_memory_ptr as usize + app_memory_size
+            app_flash.as_ptr() as usize + app_flash.len() - 1,
+            app_memory.as_ptr() as usize,
+            app_memory.as_ptr() as usize + app_memory.len() - 1
```

Comment:
I don't see why this is changed to a range including the end, rather than excluding it. Since the start and length are generally both have a large alignment, this means the newly printed address will finish with a lot of `f` characters, which I think is less readable.

Also, this change isn't applied consistently across this file, some ranges are still exclusive-ended.

If it makes thinks clearer, the format should use `..` or `..=` rather than `:` so that it's clear. But again I think the open-ended range `..` will be clearer, as it is also generally more commonly used than `..=` in Rust.

Reply:
I guess I find it more precise to use the ']' bracket in the mathematical sense to mean inclusive.

Reply:
I think in a programming / CS context, `]` unfortunately means exclusive (e.g. `"abc"[0:2]` is `"ab"` in Rust, Python; `.slice(0,2)` would be "ab" in Javascript..).

I looked up a few usages where people write memory maps. In written text, ARM seems to exclusively use inclusive ranges with `–`, i.e. `0x00000000-0x1FFFFFFF`.

I think we should maybe avoid `[]` given the ambiguity here. I'd support `..` or `–` or even `[)`

Reply:
So you are +1 on inclusive ranges, but not using `[]`?

Reply:
Thanks for mentioning Python's `[0:2]` syntax (I guess Rust moved the vertical dots into horizontal as `[0..2]` instead). I'd support:
- `0x00000000..0x20000000` (exclusive)
- `[0x00000000:0x20000000)` (exclusive)
- `[0x00000000..0x20000000)` (exclusive)
- `[0x00000000:0x20000000[` (exclusive)
- `[0x00000000..0x20000000[` (exclusive)
- `0x00000000-0x1FFFFFFF` (inclusive)
- `0x00000000..=0x1FFFFFFF` (inclusive)

with a slight preference for the exclusive end, which I find more readable for highly-aligned addresses.